File: src/main/java/com/social/meow/service/OtpService.java
package com.social.meow.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.util.Random;
import java.util.concurrent.TimeUnit;

@Service
public class OtpService {

    private final RedisTemplate<String, String> redisTemplate;
    private final SmsService smsService;
    private final Random random = new Random();

    @Value("${otp.expiration}")
    private long otpExpiration;

    public OtpService(RedisTemplate<String, String> redisTemplate, SmsService smsService) {
        this.redisTemplate = redisTemplate;
        this.smsService = smsService;
    }

    public void generateAndSendOtp(String phoneNumber) {
        String otp = String.format("%06d", random.nextInt(1000000));
        redisTemplate.opsForValue().set(getOtpKey(phoneNumber), otp, otpExpiration, TimeUnit.SECONDS);
        smsService.sendSms(phoneNumber, "Your OTP is: " + otp);
    }

    public boolean validateOtp(String phoneNumber, String otp) {
        String storedOtp = redisTemplate.opsForValue().get(getOtpKey(phoneNumber));
        return otp != null && otp.equals(storedOtp);
    }

    private String getOtpKey(String phoneNumber) {
        return "otp:" + phoneNumber;
    }
}

File: src/main/kotlin/com/social/meowkotlin/service/OtpService.kt
package com.social.meowkotlin.service

import org.springframework.beans.factory.annotation.Value
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.TimeUnit

@Service
class OtpService(private val redisTemplate: RedisTemplate<String, String>, private val smsService: SmsService) {
    private val random = Random()

    @Value("\${otp.expiration}")
    private val otpExpiration: Long = 0

    fun generateAndSendOtp(phoneNumber: String) {
        val otp = String.format("%06d", random.nextInt(1000000))
        redisTemplate.opsForValue()[getOtpKey(phoneNumber), otp, otpExpiration] = TimeUnit.SECONDS
        smsService.sendSms(phoneNumber, "Your OTP is: $otp")
    }

    fun validateOtp(phoneNumber: String, otp: String?): Boolean {
        val storedOtp = redisTemplate.opsForValue()[getOtpKey(phoneNumber)]
        return otp != null && otp == storedOtp
    }

    private fun getOtpKey(phoneNumber: String): String {
        return "otp:$phoneNumber"
    }
}
